include/uk/arch/random.h

```diff
@@ -0,0 +1,113 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) 2022, Michalis Pappas <mpappas@fastmail.fm>.
+ *                     All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef __UKARCH_RANDOM_H__
+#define __UKARCH_RANDOM_H__
+
+#ifndef __ASSEMBLY__
+#include <uk/arch/types.h>
+#endif /* !__ASSEMBLY__ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if CONFIG_UKARCH_RANDOM
+
+#include <uk/asm/random.h>
+
+/* Initialize the RNG
+ *
+ * @return 0 on success, negative value on failure
+ */
+extern int ukarch_random_init(void);
+
+/* Get a 64-bit random integer
+ *
+ * @param [in|out] val A pointer to store the generated value
+ * @return 0 on success, negative value on failure
+ */
+extern __check_result int ukarch_random_u32(__u32 *val);
+
+/* Get a 32-bit random integer
+ *
+ * @param [in|out] val A pointer to store the generated value
+ * @return 0 on success, negative value on failure
+ */
+extern __check_result int ukarch_random_u64(__u64 *val);
+
+/* Reseed the RNG and get a 64-bit random integer
+ *
+ * @param [in|out] val A pointer to store the generated value
```

Comment:
Is this really supplying a new seed value via the parameter? In that case, please fix the parameter description to make that clear.

Don't know if this is just my VSCode plugin being uncooperative, but syntax highlighting is not working for me with `|`. Do you have the same issue? I thus always use `,`.

```suggestion
 * @param [in,out] val A pointer to store the generated value
```
![image](https://user-images.githubusercontent.com/8928900/199756250-42c48a15-46da-4729-a31c-4d6d40e44d6f.png)
![image](https://user-images.githubusercontent.com/8928900/199756328-eeb2f025-b78a-4ea2-be7a-5756a605194f.png)

If the reseed does not take a value as input, then just remove the `in`.

Reply:
No, the seed is not provided by the caller. Reseeding happens by the hardware, so it's only an `[out]` parameter. Fixed.